### PR TITLE
Capsule N-Minus testing

### PR DESCRIPTION
**Pull request by jyejare**
Changes include:
- A new pytest plugin for Capsule N-minus testing that would smart collect all capsule specific tests excluding few.
  - New option of `--n-minus` for collecting only capsule specific tests
  - tests using the caspule_factory are collected
  - destructive, maintain and puppet tests are excluded since they need a separate satellite
- Introduces new setting `hostname` in capsule conf to take capsule hostname given from CI.
- **_target_capsule_host context manager now returns the cached capsule taken from CI instead of spinning of new capsule for everytime**. 
- The CI capsule is registered once/cached to CDN for RHEL contents needed by some tests at the start of the n-minus testing and would be unregistered at the end of test session.
- Since we are using single capsule already integrated with satellite in CI, we are escaping the capsule setup from fixtures for n-minus testing
- Some test amendments for nminus testing.

### CODE REVIEW

**Comment by lpramuk** [comment]
N-1 testing should be part of Upgrades, more precisely `upgrade-scenarios`

**Comment by jyejare** [comment]
@lpramuk As discussed, we need both the combinations of `Upgrade Scenarios` and `Upgrade All Tiers (Caspule specifics)` to test N-minus testing. 

So this PR take cares of `All Tier testing with old capsule`. We should be able to run existing capsule upgrade scenario without any changes though. We shall be taking care of caspule upgrade scenario in separate PR if needed !

**Review by lpramuk — CHANGES_REQUESTED** [review]
Please replace `request.config.option.n_minus` (cmdline option usage) by way more generic  `request.node.get_closest_marker(name='destructive')`

File: pytest_fixtures/core/sat_cap_factory.py [hunk 1]
```diff
@@ -148,9 +154,12 @@ def session_capsule_host(request, capsule_factory):
 
 
 @pytest.fixture
-def capsule_configured(capsule_host, target_sat):
+def capsule_configured(request, capsule_host, target_sat):
     """Configure the capsule instance with the satellite from settings.server.hostname"""
-    capsule_host.capsule_setup(sat_host=target_sat)
+    if request.config.option.n_minus:
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:159]: Nop! The new capsule creation for test case isnt just in case of destructive but for all. Hence we had to introduce a new option of nminus.

File: robottelo/hosts.py [hunk 2]
```diff
@@ -1643,6 +1643,10 @@ def capsule_setup(self, sat_host=None, capsule_cert_opts=None, **installer_kwarg
                 f'A core service is not running at capsule host\n{result.stdout}'
             )
 
+    @lru_cache
+    def cached_capsule_setup(self, sat_host):
+        self.capsule_setup(sat_host=sat_host)
```
[the remark this answers is not in the document]
Comment (jyejare) [reply robottelo/hosts.py:1648]: Yeah this would be removed! This was created as a part of testing something!

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api/
"provisioning": true

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true

File: pytest_fixtures/core/sat_cap_factory.py [hunk 3]
```diff
@@ -37,13 +38,27 @@ def _target_satellite_host(request, satellite_factory):
         yield
 
 
+@lru_cache
+def capsule_cdn_register(hostname=None):
+    cap = Capsule(hostname=hostname)
+    cap.register_to_cdn()
```
Comment (JacobCallahan) [inline pytest_fixtures/core/sat_cap_factory.py:44]: since it isn't being returned, might as well do it in one line
```suggestion
    Capsule(hostname=hostname).register_to_cdn()
```

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
One minor change

**Review by JacobCallahan — APPROVED** [review]

File: tests/foreman/api/test_capsule.py [hunk 4]
```diff
@@ -39,40 +39,51 @@ def test_positive_update_capsule(target_sat, module_capsule_configured):
...
+    except Exception as err:
+        raise err
+    finally:
+        if pytestconfig.option.n_minus:
+            capsule = target_sat.api.SmartProxy().search(query={'search': f'name = {new_name}'})
+            # Reinitializing the hostname
+            if capsule:
+                capsule = capsule[0]
+                capsule.name = module_capsule_configured.hostname
+                capsule = capsule.update(['name'])
```
Comment (ogajduse) [inline tests/foreman/api/test_capsule.py:86]: Could this part be a function defined in the test and used as a finalizer?

File: pytest_fixtures/core/sat_cap_factory.py [hunk 5]
```diff
@@ -37,13 +40,33 @@ def _target_satellite_host(request, satellite_factory):
         yield
 
 
+@lru_cache
+def capsule_cdn_register(hostname=None):
+    cap = Capsule(hostname=hostname)
+    # Enabling CDN repos on Capsule Host
+    cap.register_to_cdn()
+    # Enabling OHSnap repos on Capsule host
```
Comment (ogajduse) [inline pytest_fixtures/core/sat_cap_factory.py:48]: Could you change this statement so it matches with what is being enabled? If you look closely, none of these repositories come from Ohsnap.
https://github.com/SatelliteQE/robottelo/blob/7bdd3207149d9c7a120adae177310d8f8a086aa8/robottelo/constants/__init__.py#L254-L264

File: pytest_fixtures/core/sat_cap_factory.py [hunk 6]
```diff
@@ -37,13 +40,33 @@ def _target_satellite_host(request, satellite_factory):
...
+    if 'sanity' not in request.config.option.markexpr and not request.config.option.n_minus:
         new_cap = capsule_factory()
         yield new_cap
         new_cap.teardown()
         Broker(hosts=[new_cap]).checkin()
+    elif request.config.option.n_minus:
+        if not settings.capsule.hostname:
+            hosts = Capsule.get_hosts_from_inventory(filter="'cap' in @inv.name")
+            settings.capsule.hostname = hosts[0].hostname
+        caps = Capsule(hostname=settings.capsule.hostname)
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py]: Yup! Updated that during TFA in 6.14 branch https://github.com/jyejare/robottelo/tree/614_Capsule_n_minus, even now I pushed it here!

File: pytest_fixtures/core/sat_cap_factory.py [hunk 7]
```diff
@@ -37,13 +40,33 @@ def _target_satellite_host(request, satellite_factory):
...
-    if 'sanity' not in request.config.option.markexpr:
+    if 'sanity' not in request.config.option.markexpr and not request.config.option.n_minus:
         new_cap = capsule_factory()
         yield new_cap
         new_cap.teardown()
         Broker(hosts=[new_cap]).checkin()
+    elif request.config.option.n_minus:
+        if not settings.capsule.hostname:
+            hosts = Capsule.get_hosts_from_inventory(filter="'cap' in @inv.name")
+            settings.capsule.hostname = hosts[0].hostname
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:66]: We are in CI are spinning only one pair of Satellite and Capsule, hence there will always be only one satellite and capsule. 

For local setting one will always pass capsule hostname in capsule.yaml conf !

File: pytest_fixtures/core/sat_cap_factory.py [hunk 8]
```diff
@@ -37,13 +40,33 @@ def _target_satellite_host(request, satellite_factory):
         yield
 
 
+@lru_cache
+def capsule_cdn_register(hostname=None):
+    cap = Capsule(hostname=hostname)
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:45]: Yup! Updated that during TFA in 6.14 branch https://github.com/jyejare/robottelo/tree/614_Capsule_n_minus, even now I pushed it here!

File: robottelo/hosts.py [hunk 9]
```diff
@@ -1684,7 +1684,10 @@ def run_installer_arg(self, *args, timeout='20m'):
             timeout=timeout,
         )
         if result.status != 0:
-            raise SatelliteHostError(f'Failed to execute with argument: {result.stderr}')
+            raise SatelliteHostError(
+                f'Failed to execute with arguments: {installer_args} and,'
+                f' the stdout is {result.stdout}'
+            )
```
[the remark this answers is not in the document]
Comment (jyejare) [reply robottelo/hosts.py:1690]: The earlier statement was :
`Failed to execute with **argument**: {result.stderr}` so it should have shown the arguments where as I dont know why it was showing the stderr.

Now I added additinal statement to see the stdout, I can change it to see the error though.

File: tests/foreman/api/test_capsule.py (hunk 4 above)
Comment (jyejare) [reply tests/foreman/api/test_capsule.py:86]: Good Suggestion! Almost forgot about finalizer. Done now!

File: pytest_fixtures/core/sat_cap_factory.py (hunk 5 above)
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py]: Yup! Updated that during TFA in 6.14 branch https://github.com/jyejare/robottelo/tree/614_Capsule_n_minus, even now I pushed it here!

File: pytest_fixtures/core/sat_cap_factory.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ogajduse) [reply pytest_fixtures/core/sat_cap_factory.py:57]: Got it. If I understand it correctly, this is something that could be considered as an improvement for the future work on capsule n minus.

[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:57]: Yes, To improvise the time or concurrency we could probably go for team specific jobs in CI OR multiple capsules where test has to be mapped with the particular capsule so that the entire test would run on same capsule and not split into multiple.

BTW, that is not yet the need in near future, if its needed we will reevaluate.

**Comment by jyejare** [comment]
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: '6.15'

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: '6.15'
```

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: '6.15'
    ROBOTTELO_server__version__release: '6.15'
```

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: "'6.15.0'"
    ROBOTTELO_server__version__release: "'6.15.0'"
```

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: '6.15.0'
    ROBOTTELO_server__version__release: '6.15.0'
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 37
Build Status: UNSTABLE
PRT Comment: pytest --uses-fixtures capsule_factory tests/foreman/api
Test Result : ========= 6 skipped, 2764 deselected, 13 warnings, 33 errors in 47.32s =========
```

**Comment by jyejare** [comment]
```
trigger: test-robottelo
pytest: --uses-fixtures capsule_factory tests/foreman/api
provisioning: true
env:
    ROBOTTELO_server__deploy_arguments__deploy_sat_version: '6.15.0'
    ROBOTTELO_server__deploy_arguments__deploy_snap_version: ''
    ROBOTTELO_server__version__release: '6.15.0'
    ROBOTTELO_server__version__snap: ''
```

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 5956
Build Status: UNSTABLE
PRT Comment: pytest --uses-fixtures capsule_factory tests/foreman/api --external-logging --include-onprem-provisioning
Test Result : = 3 failed, 28 passed, 6 skipped, 2764 deselected, 2602 warnings, 2 errors in 13240.51s (3:40:40) =
```

**Comment by jyejare** [comment]
The test failures here are unrelated to the framework changes / test changes made in this PR!  

We should be good to get this merged now, the counterpart of this PR is merged in CI !

CC: @JacobCallahan @ogajduse @SatelliteQE/airgun-tier-2-reviewers

File: robottelo/hosts.py [hunk 10]
```diff
@@ -1599,6 +1599,21 @@ def get_features(self):
         """Get capsule features"""
         return requests.get(f'https://{self.hostname}:9090/features', verify=False).text
 
+    def enable_capsule_downstream_repos(self):
+        """Enabling CDN repos and capsule downstream repos on Capsule Host"""
```
Comment (ogajduse) [inline robottelo/hosts.py:1603]:
```suggestion
        """Enable CDN repos and capsule downstream repos on Capsule Host"""
```

File: conf/capsule.yaml.template [hunk 11]
```diff
@@ -1,4 +1,6 @@
 CAPSULE:
+  # Capsule hostname for N-minus testing
```
Comment (Gauravtalreja1) [inline conf/capsule.yaml.template:2]: hostname can be generally used, so comment can be removed?, and I don't see this setting added on CI conf, instead it is directly used in pipeline, so do we need it here?
```suggestion

```

File: conftest.py [hunk 12]
```diff
@@ -21,6 +21,7 @@
     'pytest_plugins.requirements.update_requirements',
     'pytest_plugins.sanity_plugin',
     'pytest_plugins.video_cleanup',
+    'pytest_plugins.capsule_n-minus',
```
Comment (Gauravtalreja1) [inline conftest.py:24]: To simplify this naming?
```suggestion
    'pytest_plugins.capsule_nminus',
```

File: pytest_fixtures/core/sat_cap_factory.py [hunk 13]
```diff
@@ -176,21 +194,23 @@ def large_capsule_configured(large_capsule_host, target_sat):
 
 
 @pytest.fixture(scope='module')
-def module_capsule_configured(module_capsule_host, module_target_sat):
+def module_capsule_configured(request, module_capsule_host, module_target_sat):
     """Configure the capsule instance with the satellite from settings.server.hostname"""
-    module_capsule_host.capsule_setup(sat_host=module_target_sat)
+    if not request.config.option.n_minus:
```
Comment (Gauravtalreja1) [inline pytest_fixtures/core/sat_cap_factory.py:199]: Could you explain what this check is about?

File: pytest_fixtures/core/sat_cap_factory.py [hunk 14]
```diff
@@ -37,13 +38,29 @@ def _target_satellite_host(request, satellite_factory):
...
         Broker(hosts=[new_cap]).checkin()
+    elif request.config.option.n_minus:
+        if not settings.capsule.hostname:
+            hosts = Capsule.get_hosts_from_inventory(filter="'cap' in @inv.name")
+            settings.capsule.hostname = hosts[0].hostname
+            cap = hosts[0]
+        else:
+            cap = Capsule.get_host_by_hostname(settings.capsule.hostname)
+        # Capsule needs RHEL contents for some tests
+        cached_capsule_cdn_register(hostname=settings.capsule.hostname)
```
Comment (Gauravtalreja1) [inline pytest_fixtures/core/sat_cap_factory.py:62]: why capsule registration to CDN and repo enable required to be cached here?

File: pytest_plugins/capsule_n-minus.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        selected.append(item)
+
+    config.hook.pytest_deselected(items=deselected)
+    items[:] = selected
+
+
+def pytest_sessionfinish(session, exitstatus):
+    # Unregister the capsule from CDN after all tests
+    if session.config.option.n_minus and not session.config.option.collectonly:
+        caps = Capsule(hostname=settings.capsule.hostname)
```
[the remark this answers is not in the document]
Comment (Gauravtalreja1) [reply pytest_plugins/capsule_n-minus.py:54]: `Capsule.get_host_by_hostname(hostname=settings.capsule.hostname).unregister()` ?

File: robottelo/hosts.py [hunk 16]
```diff
@@ -1599,6 +1599,21 @@ def get_features(self):
...
+        # CDN Repos
+        self.register_to_cdn()
+        for repo in getattr(constants, f"OHSNAP_RHEL{self.os_version.major}_REPOS"):
+            result = self.enable_repo(repo, force=True)
+            if result.status:
+                raise CapsuleHostError(f'Repo enable at capsule host failed\n{result.stdout}')
```
Comment (Gauravtalreja1) [inline robottelo/hosts.py:1609]: When you register to cdn, this repos are enabled as default so we won't need additional step to enable this rhel repos
```suggestion

```

File: tests/foreman/api/test_capsulecontent.py [hunk 17]
```diff
@@ -1312,8 +1313,9 @@ def test_positive_remove_capsule_orphans(
             data={'environment_id': function_lce_library.id}
         )
         result = capsule_configured.nailgun_capsule.content_lifecycle_environments()
-        assert len(result['results']) == 1
-        assert result['results'][0]['id'] == function_lce_library.id
+        if not pytestconfig.option.n_minus:
```
Comment (vsedmik) [inline tests/foreman/api/test_capsulecontent.py:1316]: @jyejare, do I guess correctly the need for this `if` is the fact that n_minus capsule is session-scoped like and thus contains more LCEs?

This test case is supposed to use the function-scoped `capsule_configured` (not the module scoped one) for the reason of verification that all RPM artifacts are cleared after orphan cleanup. This will not happen for a session-scoped capsule, since any previously run test (which syncs at least one yum repo) will leave RPM artifacts on the capsule and the test will fail on L1351.

In other words, I don't think this test case is good for n_minus scenario.

Comment (jyejare) [reply tests/foreman/api/test_capsulecontent.py:1316]: Hello @vsedmik for comment, Yes the capsule here is a single capsule and given by CI where Satellite is at higher version and capsule at lower (We call this as n-minus).

Since the capsule has all the data from all the tests (like LCEs in this case), we had to skip some assertions like these. Apart from these assertions rest assertions are being validated. But As per the design of n-minus setup we cant have on-demand capsules like the scoped caspules once etc.

Since this is a smart test collection (that is without markers) , to uncollect some test we have to introduce some uncollect marker OR keep the test with custom assertions. I would vote for second.

File: pytest_fixtures/core/sat_cap_factory.py [hunk 18]
```diff
@@ -201,7 +221,9 @@ def module_capsule_configured_mqtt(module_capsule_configured):
     result = module_capsule_configured.execute('firewall-cmd --permanent --add-port="1883/tcp"')
     assert result.status == 0, 'Failed to open mqtt port on capsule'
     module_capsule_configured.execute('firewall-cmd --reload')
-    return module_capsule_configured
+    yield module_capsule_configured
+    if request.config.option.n_minus:
+        raise TypeError('The teardown is missed for MQTT configuration undo for nminus testing')
```
[the remark this answers is not in the document]
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:226]: TBH We dont need this change but since the setup here without teardown could have impact on feature tests since they are running on same satellite.

Since I am unaware of how the best teardown should happen, I am leaving upto Teams to do it by adding this error. :)

Its a different trick, but the expectation is the person doing the TFA should resolve it !

File: tests/foreman/api/test_capsule.py [hunk 19]
```diff
@@ -74,6 +74,16 @@ def test_positive_update_capsule(target_sat, module_capsule_configured):
     assert capsule.url in [cps.url for cps in capsules]
     assert capsule.name in [cps.name for cps in capsules]
 
+    @request.addfinalizer
+    def _finalize():
+        if pytestconfig.option.n_minus:
+            cap = target_sat.api.SmartProxy().search(query={'search': f'name = {new_name}'})
+            # Reinitializing the hostname
+            if cap:
```
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/api/test_capsule.py:82]: Updated! But we should find/write the new ruff rule for this for auto-formatting!

File: conf/capsule.yaml.template (hunk 11 above)
Comment (jyejare) [reply conf/capsule.yaml.template:2]: Since all conf templates have some information about setting, I added this. Keeping it is no harm.

Its good to keep setting in sync in CI and robottelo especially in confs, so that local execution of tests wont fail.

I will open an MR in jenkins to create the conf.

File: conftest.py (hunk 12 above)
Comment (jyejare) [reply conftest.py:24]: May be newbie wont understand it quick. Also there is no standard in naming convention. Skipping it as non-blocking.

File: pytest_fixtures/core/sat_cap_factory.py (hunk 13 above)
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:199]: Here, If the `--n-minus` option is given to pytest command for collecting / executing all capsule tests, but since the capsule is already configured with satellite and passed in from capsule conf, we dont need to reconfigure the capsule. 

The condition here skips the capsule configuration for n-minus testing!

File: pytest_fixtures/core/sat_cap_factory.py (hunk 14 above)
Comment (jyejare) [reply pytest_fixtures/core/sat_cap_factory.py:62]: Since we are using single capsule, and feeded from CI for n-minus testing, we dont need to register it again and again.

File: robottelo/hosts.py (hunk 16 above)
Comment (jyejare) [reply robottelo/hosts.py:1609]: Agree and I observed the same! But there are existing functions in robottelo which are trying to do it explicitly hence I had to repeat here. 

If we want to remove this, we will remove from all occurrences of it in a separate PR.

File: pytest_plugins/capsule_n-minus.py [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        is_destructive = item.get_closest_marker('destructive')
+        # Deselect Destructive tests and tests without capsule_factory fixture
+        if 'capsule_factory' not in item.fixturenames or is_destructive:
+            deselected.append(item)
+            continue
+        # Ignoring all puppet tests as they are destructive in nature
+        # and needs its own satellite for verification
+        if 'session_puppet_enabled_sat' in item.fixturenames:
+            deselected.append(item)
+            continue
```
Comment (Gauravtalreja1) [inline pytest_plugins/capsule_n-minus.py:39]: For tests using `session_puppet_enabled_sat` there's a capsule configured from fixture  `session_puppet_enabled_capsule`, so shouldn't it be used for n-minus instead of ignoring it?

File: tests/foreman/api/test_capsule.py [hunk 21]
```diff
@@ -74,6 +74,15 @@ def test_positive_update_capsule(target_sat, module_capsule_configured):
...
 
+    @request.addfinalizer
+    def _finalize():
+        if pytestconfig.option.n_minus:
+            # Reinitializing the hostname
+            if cap := target_sat.api.SmartProxy().search(query={'search': f'name = {new_name}'}):
+                cap = cap[0]
+                cap.name = module_capsule_configured.hostname
+                cap.update(['name'])
+
```
Comment (Gauravtalreja1) [inline tests/foreman/api/test_capsule.py:85]: Can we move this finalizer method to top, if in case this test fails this teardown could get skip as its in end, right?

File: pytest_plugins/capsule_n-minus.py (hunk 20 above)
Comment (jyejare) [reply pytest_plugins/capsule_n-minus.py:39]: Puppet test cases need there own Satellite (and they are destructive to be used with existing satellite) and since we cant provide new satellite in n-minus setup, its good to skip!

File: tests/foreman/api/test_capsulecontent.py (hunk 17 above)
Comment (jyejare) [reply tests/foreman/api/test_capsulecontent.py:1316]: @vsedmik Atleast, I dont see the test failing after I removed these 2 assertions, but If you feel like we should be just skipping it, I can do it!

Comment (vsedmik) [reply tests/foreman/api/test_capsulecontent.py]: @jyejare Hmm, does it pass when run standalone or within the whole test suite? Do you know the test order run before this one?

Comment (jyejare) [reply tests/foreman/api/test_capsulecontent.py]: @vsedmik AFAIR, it passes and passes within the whole suite. I can PM you the order because the list would be big here to paste.

**Review by Gauravtalreja1 — APPROVED** [review]
